crates/consensus-logic/src/asm_worker_context.rs

```
@@ -116,36 +117,46 @@ impl WorkerContext for AsmWorkerCtx {
-        self.mmr_manager
+    fn append_manifest_to_mmr(&self, manifest_hash: Hash) -> WorkerResult<u64> {
+        self.mmr_handle
             .append_leaf_blocking(manifest_hash)
             .map_err(|e| {
-                tracing::error!(?e, "Failed to append leaf to MMR");
+                error!(?e, "Failed to append leaf to MMR");
                 WorkerError::DbError
             })
     }
 
-    fn store_manifest_hash(&self, index: u64, hash: [u8; 32]) -> WorkerResult<()> {
-        self.asmman
-            .store_manifest_hash(index, Buf32(hash))
-            .map_err(conv_db_err)
-    }
-
     fn generate_mmr_proof(&self, index: u64) -> WorkerResult<strata_merkle::MerkleProofB32> {
-        self.mmr_manager.generate_proof(index).map_err(|e| {
-            tracing::error!(?e, index, "Failed to generate MMR proof");
+        self.mmr_handle.generate_proof(index).map_err(|e| {
+            error!(?e, index, "Failed to generate MMR proof");
             WorkerError::MmrProofFailed { index }
         })
     }
 
-    fn get_manifest_hash(&self, index: u64) -> WorkerResult<Option<[u8; 32]>> {
-        self.asmman
-            .get_manifest_hash(index)
-            .map(|opt| opt.map(|buf| buf.0))
-            .map_err(conv_db_err)
+    fn get_manifest_hash(&self, index: u64) -> WorkerResult<Option<Hash>> {
+        let num_leaves = self.mmr_handle.num_leaves_blocking().map_err(|e| {
+            error!(?e, "Failed to get MMR num_leaves");
+            WorkerError::DbError
+        })?;
+
+        if index >= num_leaves {
+            return Err(WorkerError::ManifestIndexOutOfBound {
+                index,
+                max: num_leaves,
+            });
+        }
+
+        self.mmr_handle
+            .get_leaf_hash_blocking(index)
+            .map(Some)
+            .map_err(|e| {
+                error!(?e, index, "Failed to get leaf hash from MMR");
+                WorkerError::DbError
+            })
     }
 }
```

A few thoughts here.

1. Since we check the error condition everytime, should we directly return `WorkerResult<Hash>`? instead of `WorkerResult<Option<Hash>>`?
2. Do we need to log the errors here? Otherwise we could just propagate the errors to the caller.

CC: @evgenyzdanovich